src/main/java/ee/tuleva/onboarding/mandate/batch/MandateBatchService.java

```diff
@@ -0,0 +1,67 @@
+package ee.tuleva.onboarding.mandate.batch;
+
+import static java.util.stream.Collectors.toList;
+
+import ee.tuleva.onboarding.auth.principal.AuthenticatedPerson;
+import ee.tuleva.onboarding.mandate.MandateFileService;
+import ee.tuleva.onboarding.mandate.generic.GenericMandateService;
+import ee.tuleva.onboarding.mandate.signature.SignatureFile;
+import ee.tuleva.onboarding.user.User;
+import java.util.List;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class MandateBatchService {
+  private final MandateBatchRepository mandateBatchRepository;
+  private final MandateFileService mandateFileService;
+  private final GenericMandateService genericMandateService;
+
+  public Optional<MandateBatch> getByIdAndUser(Long id, User user) {
+    var batch =
+        mandateBatchRepository
+            .findById(id)
+            .filter(
+                mandateBatch ->
+                    mandateBatch.getMandates().stream()
+                        .allMatch(mandate -> mandate.getUser().equals(user)));
+
+    if (batch.isEmpty()) {
+      log.error("Invalid user {} for mandate batch {}", user.getId(), id);
+    }
+
+    return batch;
+  }
+
+  public MandateBatch createMandateBatch(
+      AuthenticatedPerson authenticatedPerson, MandateBatchDto mandateBatchDto) {
+
+    var mandateBatch = MandateBatch.builder().status(MandateBatchStatus.INITIALIZED).build();
+
+    var savedMandateBatch = mandateBatchRepository.save(mandateBatch);
+    var mandates =
+        mandateBatchDto.getMandates().stream()
+            .map(
+                mandateDto ->
+                    genericMandateService.createGenericMandate(
+                        authenticatedPerson, mandateDto, mandateBatch))
+            .collect(toList());
```

Comment:
Use immutable `.toList()`, if possible

Reply:
We had a problem with Hibernate with immutable `toList` here